File: scaly/src/containers/stringbuilder.rs
use containers::{Array, Ref, String};
use memory::page::Page;

#[derive(Copy, Clone)]
pub struct StringBuilder {
    buffer: Array<u8>,
}

impl StringBuilder {
    pub fn new(_rp: *mut Page) -> Ref<StringBuilder> {
        Ref::new(
            _rp,
            StringBuilder {
                buffer: Array::new(),
            },
        )
    }

    pub fn from_character(_rp: *mut Page, character: char) -> Ref<StringBuilder> {
        let mut string_builder = StringBuilder::new(_rp);
        string_builder.append_character(character);
        string_builder
    }

    pub fn append_character(&mut self, character: char) {
        self.buffer.add(character as u8);
    }

    pub fn get_length(&self) -> usize {
        return self.buffer.get_length();
    }

    pub fn to_string(&self, _rp: *mut Page) -> String {
        String::create(_rp, self.buffer.get_buffer(), self.buffer.get_length())
    }
}

File: scaly/containers/StringBuilder.cpp
#include "Containers.cpp"

namespace scaly {
namespace containers {

using namespace scaly::memory;

    StringBuilder::StringBuilder(Array<char> _buffer)
      : buffer(_buffer) {}

    StringBuilder::StringBuilder() : StringBuilder(Array<char>()) {}

    StringBuilder::StringBuilder(char character) : StringBuilder() {
        this->append(character);
    }

    StringBuilder::StringBuilder(const String& string) : StringBuilder() {
        this->append(string);
    }

    StringBuilder::StringBuilder(const char* c_string) : StringBuilder() {
        this->append(c_string);
    }

    void StringBuilder::append(char character) {
        this->buffer.add(character);
    }

    void StringBuilder::append(const String& string) {
        this->buffer.add(Vector<char>(string.get_buffer(), string.get_length()));
    }

    void StringBuilder::append(const char* c_string) {
        auto length = strlen(c_string);
        if (length == 0)
            return;

        this->buffer.add(Vector<char>((char*)c_string, strlen(c_string)));
    }

    size_t StringBuilder::get_length() {
        return this->buffer.get_length();
    }

    String StringBuilder::to_string(Page*_rp) {
        return String(_rp, this->buffer.get_buffer(), this->buffer.get_length());
    }

}
}
